### PR TITLE
chore(test): add e2e-test-containers

### PR DESCRIPTION
**Pull request by joshuachp**
Add a binary to simulate or test the reception of DeviceEvents from
Astarte for the edgehog-device-runtime-containers.

Supersedes #316

### CODE REVIEW

File: e2e-test-containers/src/cli.rs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+}
+
+#[derive(Debug, Clone, Subcommand)]
+pub enum Command {
+    /// Send the data to astarte.
+    Send {
+        /// Token with access to app engine API to send data to Astarte.
+        #[arg(long, env = "ASTARTE_TOKEN")]
+        token: String,
+        /// Token with access to app engine API to send data to Astarte.
```
Comment (rgallor) [inline e2e-test-containers/src/cli.rs:39]: doc comment already used

File: e2e-test-containers/src/cli.rs [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        #[arg(long, env = "ASTARTE_API_URL")]
+        appengine_url: String,
+        #[arg(long, default_value = "false")]
+        curl: bool,
```
Comment (rgallor) [inline e2e-test-containers/src/cli.rs:43]: I'd add a comment to this field for completeness

File: e2e-test-containers/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+)> {
+    let mut mqtt_config = MqttConfig::new(
+        &astarte.realm,
+        &astarte.device_id,
+        Credential::secret(astarte.credentials_secret.clone()),
+        &astarte.pairing_url,
+    );
+    mqtt_config.ignore_ssl_errors();
+
+    // 3. Create the device instance
```
Comment (rgallor) [inline e2e-test-containers/src/main.rs:52]: I think you could remove this comment

File: e2e-test-containers/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+                client.print_curl(data).await?;
+            } else {
+                client.read(data).await?;
+            }
+        }
+        cli::Command::Receive => {
+            let (client, connection_handle) = connect(&cli.astarte).await?;
+
+            let recv_handle =
+                tokio::spawn(async move { receive(client, &cli.astarte.store_dir).await });
```
Comment (rgallor) [inline e2e-test-containers/src/main.rs:101]: You could use a `JoinSet` and pass it to the connect method and after receiving a ctrl_c signal abort the tasks in the joinset

Comment (rgallor) [reply e2e-test-containers/src/main.rs:101]: Still, the code is correct and it is used for tests, so feel free to ignore this comment

**Review by rgallor — CHANGES_REQUESTED** [review]

File: e2e-test-containers/src/main.rs (hunk 4 above)
Comment (joshuachp) [reply e2e-test-containers/src/main.rs:101]: I abort the tasks in order, the recv before of the connection

File: e2e-test-containers/src/main.rs [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+}
+
+#[tokio::main]
+async fn main() -> color_eyre::Result<()> {
+    let cli = Cli::parse();
+
+    color_eyre::install()?;
+
+    let filter = EnvFilter::builder()
+        .with_default_directive("edgehog_device_runtime_docker=DEBUG".parse()?)
```
Comment (harlem88) [inline e2e-test-containers/src/main.rs:73]:
```suggestion
        .with_default_directive("edgehog_device_runtime_containers=DEBUG".parse()?)
```

**Review by harlem88 — CHANGES_REQUESTED** [review]

File: .reuse/dep5 [hunk 6]
```diff
@@ -3,5 +3,10 @@ Upstream-Name: edgehog-device-runtime
 Source: https://github.com/edgehog-device-manager/edgehog-device-runtime
 
 Files: CHANGELOG.md edgehog-device-runtime-containers/CHANGELOG.md
-Copyright: 2022 SECO Mind Srl
+Copyright: 2022-2024sSECO Mind Srl
```
[the remark this answers is not in the document]
Comment (joshuachp) [reply .reuse/dep5:6]: I also converted the file to the new REUSE.toml format

**Review by harlem88 — APPROVED** [review]

**Review by rgallor — APPROVED** [review]

**Pull request merged**